scipio/src/controllers/deadline_queue.rs

```diff
@@ -0,0 +1,470 @@
+    /// starts.
+    fn deadline(&self) -> Duration;
+
+    /// The action to execute. Usually your struct will implement an async function that you want
+    /// to see completed at a particular deadline, and then the implementation of this would be:
+    ///
+    /// ```ignore
+    /// fn action(&self) -> Pin<Box<dyn Future<Output = io::Result<Duration>> + '_>> {
+    ///    Box::pin(self.my_action())
+    /// }
+    /// ```
+    fn action(&self) -> Pin<Box<dyn Future<Output = io::Result<Self::Output>> + '_>>;
+
+    /// The total amount of units to be processed.
+    ///
+    /// This could be anything you want:
+    /// * If you are flushing a file, this could indicate the size in bytes of the buffer
+    /// * If you are scanning an array, this could be the number of elements.
+    ///
+    /// As long as this quantity is consistent with [`processed_units`] the controllers should work.
+    ///
+    /// This need not be static. On the contrary: as you are filling a new buffer you can already
+    /// add it to the queue and increase its total units as the buffer is written to. This can
+    /// lead to smoother operation as opposed to just adding a lot of units at once.
+    ///
+    /// [`processed_units`]: trait.DeadlineSource.html#method.processed_units
+    fn total_units(&self) -> u64;
+
+    /// The amount of units that were already processed.
+    ///
+    /// The units should match the quantities specified in [`total_units`].
+    /// The more often the system is made aware of processed units, the smoother the
+    /// controller will be.
+    ///
+    /// For example, you can buffer all updates and just inform that processed_units == total_units
+    /// at the end of the process, but then the controller would be a step function.
+    ///
+    /// [`total_units`]: trait.DeadlineSource.html#method.total_units
+    fn processed_units(&self) -> u64;
+}
+
+impl<T> fmt::Debug for dyn DeadlineSource<Output = T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "DeadlineSource processed {} out of {}",
+            self.processed_units(),
+            self.total_units()
+        )
+    }
+}
+
+#[derive(Debug)]
+/// Allows the priority of the [`DeadlineQueue`] to be temporarily bumped.
+///
+/// The priority is bumped for as long as this object is alive. This is useful
+/// in situations where, despite having a deadline we may never want the priority
+/// to fall too low.
+///
+/// This could be because a user started watching the process, a shutdown sequence
+/// was initiated, etc.
+///
+/// [`DeadlineQueue`]: struct.DeadlineQueue.html
+pub struct PriorityBump<T> {
+    queue: Rc<InnerQueue<T>>,
+}
+
+impl<T> PriorityBump<T> {
+    fn new(queue: Rc<InnerQueue<T>>) -> PriorityBump<T> {
+        queue.min_shares.replace(250);
+        PriorityBump { queue }
+    }
+}
+
+impl<T> Drop for PriorityBump<T> {
+    fn drop(&mut self) {
+        self.queue.min_shares.replace(1);
```

Comment:
I guess no one will do 

```ts
let pb = PriorityBump::new(q)
let pb2 = PriorityBump::new(pb.queue)
drop(pb) // back to 1
```

Reply:
I would consider this a bug.
Notice how you can't even specify the priority you want in the API, so this is not like a verbose flag that you can pass many.